From ea83ed14168f681e0ac9ee140437abd109a23814 Mon Sep 17 00:00:00 2001
From: Don Zickus <dzickus@redhat.com>
Date: Tue, 16 Sep 2014 17:07:22 -0400
Subject: [usb] cdc-acm: fix double usb_autopm_put_interface() in acm_port_activate()

Message-id: <1410887463-20674-371-git-send-email-dzickus@redhat.com>
Patchwork-id: 95636
O-Subject: [RHEL7 PATCH 370/591] USB: cdc-acm: fix double usb_autopm_put_interface() in acm_port_activate()
Bugzilla: 1110939
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>

https://bugzilla.redhat.com/show_bug.cgi?id=1110939

commit 070c0b17f6a1ba39dff9be112218127e7e8fd456
Author: Alexey Khoroshilov <khoroshilov@ispras.ru>
Date:   Sat Apr 12 02:10:45 2014 +0400

    USB: cdc-acm: fix double usb_autopm_put_interface() in acm_port_activate()

    If acm_submit_read_urbs() fails in acm_port_activate(), error handling
    code calls usb_autopm_put_interface() while it is already called
    before acm_submit_read_urbs(). The patch reorganizes error handling code
    to avoid double decrement of USB interface's PM-usage counter.

    Found by Linux Driver Verification project (linuxtesting.org).

    Signed-off-by: Alexey Khoroshilov <khoroshilov@ispras.ru>
    Acked-by: Oliver Neukum <oliver@neukum.org>
    Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/usb/class/cdc-acm.c b/drivers/usb/class/cdc-acm.c
index 900f7ff..d5d2c92 100644
--- a/drivers/usb/class/cdc-acm.c
+++ b/drivers/usb/class/cdc-acm.c
@@ -518,13 +518,16 @@ static int acm_port_activate(struct tty_port *port, struct tty_struct *tty)
  if (usb_submit_urb(acm->ctrlurb, GFP_KERNEL)) {
   dev_err(&acm->control->dev,
    "%s - usb_submit_urb(ctrl irq) failed\n", __func__);
+  usb_autopm_put_interface(acm->control);
   goto error_submit_urb;
  }
 
  acm->ctrlout = ACM_CTRL_DTR | ACM_CTRL_RTS;
  if (acm_set_control(acm, acm->ctrlout) < 0 &&
-     (acm->ctrl_caps & USB_CDC_CAP_LINE))
+     (acm->ctrl_caps & USB_CDC_CAP_LINE)) {
+  usb_autopm_put_interface(acm->control);
   goto error_set_control;
+ }
 
  usb_autopm_put_interface(acm->control);
 
@@ -549,7 +552,6 @@ error_submit_read_urbs:
 error_set_control:
  usb_kill_urb(acm->ctrlurb);
 error_submit_urb:
- usb_autopm_put_interface(acm->control);
 error_get_interface:
 disconnected:
  mutex_unlock(&acm->mutex);
-- 
1.7.1